### PR TITLE
Allows the ability to configure bulkheads using config properties

### LINKED ISSUE #224: Confusing `enableSemaphoreDefaultBulkhead` property
I am not sure if this is a bug, but It feels like one.

**Background**
As we know, by default, if no Bulkhead is configured, then `Spring Cloud Circuit Breaker` will use the `ThreadPoolBulkhead` for each circuit breaker. For us, we prefer semaphores for our event driven services, we have no need for thread pools. So, we found the property `spring.cloud.circuitbreaker.resilience4j.enableSemaphoreDefaultBulkhead` and we set it to `true`.

**Confusion**
For a service where all HTTP clients use a semaphore bulkhead, except one. What happens if the above property is set to `true` and `ThreadPoolBulkhead` is configured for the HTTP client `X`? The answer is a semaphore bulkhead!

This is the confusion, a `ThreadPoolBulkhead` configuration will never bite when the `enableSemaphoreDefaultBulkhead` is set to `true`. This property should instead be named `forceSemaphoreBulkhead`.

The code seems incorrect in my opinion. This parameter should `default` to a semaphore bulkhead, if no configuration says otherwise. The code I am talking about resides in the class `Resilience4jBulkheadProvider`.

```
	private boolean useSemaphoreBulkhead(String id) {
		return semaphoreDefaultBulkhead
				|| (bulkheadRegistry.find(id).isPresent() && threadPoolBulkheadRegistry.find(id).isEmpty());
	}
```

So basically the `semaphoreDefaultBulkhead`, in my opinion, should be checked if no `SemaphoreBulkhead` or `ThreadPoolBulkhead` exists, then default to a `SemaphoreBulkhead`, else a `ThreadPoolBulkhead`.

**Why even Mix?**
Well, the `ThreadPoolBulkhead` is preferable for our non-event driven services. Semaphores cannot be queued, only timed. But this applies only for certain clients. As it is now, we need to turn this flag off, and we have to explicitly configure all clients. 

Have we misunderstood this flag?

Thank you

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Also addresses issue where semaphore bulkhead was used when threadpool config was supplied.

Fixes #224

### CODE REVIEW

**Comment by ryanjbaxter** [comment]
@OathMeadow this PR addresses the issues in #224.  Let me know what you think.

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4jBulkheadProvider.java [hunk 1]
```diff
@@ -134,10 +141,20 @@ private <T> Supplier<CompletionStage<T>> decorateBulkhead(final String id, final
...
 	}
 
+	private Resilience4jBulkheadConfigurationBuilder.BulkheadConfiguration getConfiguration(String id) {
+		Resilience4jBulkheadConfigurationBuilder builder = new Resilience4jBulkheadConfigurationBuilder();
+		Optional<BulkheadConfig> bulkheadConfiguration = bulkheadRegistry.getConfiguration(id);
+		Optional<ThreadPoolBulkheadConfig> threadPoolBulkheadConfig = threadPoolBulkheadRegistry.getConfiguration(id);
+		builder.bulkheadConfig(bulkheadConfiguration.orElse(bulkheadRegistry.getDefaultConfig()));
+		builder
+			.threadPoolBulkheadConfig(threadPoolBulkheadConfig.orElse(threadPoolBulkheadRegistry.getDefaultConfig()));
+		return builder.build();
```
Comment (OathMeadow) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4jBulkheadProvider.java:151]: I was wondering what function the `defaultConfiguration` would serve.
Currently only used in `addBulkheadCustomizer` and `addThreadPoolBulkheadCustomizer`.

We could maybe handle this similair to what we do in `Resilience4JCircuitBreakerFactory`.

```
	private Resilience4JCircuitBreaker create(String id, String groupName,
			ExecutorService circuitBreakerExecutorService) {
		Resilience4JConfigBuilder.Resilience4JCircuitBreakerConfiguration defaultConfig = getConfigurations()
				.computeIfAbsent(id, defaultConfiguration);
		CircuitBreakerConfig circuitBreakerConfig = this.circuitBreakerRegistry.getConfiguration(id)
				.orElseGet(() -> this.circuitBreakerRegistry.getConfiguration(groupName)
						.orElseGet(defaultConfig::getCircuitBreakerConfig));
		TimeLimiterConfig timeLimiterConfig = this.timeLimiterRegistry.getConfiguration(id)
				.orElseGet(() -> this.timeLimiterRegistry.getConfiguration(groupName)
						.orElseGet(defaultConfig::getTimeLimiterConfig));
		if (resilience4JConfigurationProperties.isDisableThreadPool()) {
			return new Resilience4JCircuitBreaker(id, groupName, circuitBreakerConfig, timeLimiterConfig,
					circuitBreakerRegistry, timeLimiterRegistry, Optional.ofNullable(circuitBreakerCustomizers.get(id)),
					bulkheadProvider);
		}
		else {
			return new Resilience4JCircuitBreaker(id, groupName, circuitBreakerConfig, timeLimiterConfig,
					circuitBreakerRegistry, timeLimiterRegistry, circuitBreakerExecutorService,
					Optional.ofNullable(circuitBreakerCustomizers.get(id)), bulkheadProvider,
					this.resilience4JConfigurationProperties.isDisableTimeLimiter());
		}

	}
```

File: spring-cloud-circuitbreaker-resilience4j/src/test/java/org/springframework/cloud/circuitbreaker/resilience4j/BulkheadConfigurationTests.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.test.context.runner.WebApplicationContextRunner;
+import org.springframework.context.annotation.Configuration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Ryan Baxter
+ */
+public class BulkheadConfigurationTests {
```
Comment (OathMeadow) [inline spring-cloud-circuitbreaker-resilience4j/src/test/java/org/springframework/cloud/circuitbreaker/resilience4j/BulkheadConfigurationTests.java:34]: Maybe we should also add default config tests.

example: `resilience4j.threadPoolBulkHead.configs.default.queueCapacity=30`

File: spring-cloud-circuitbreaker-resilience4j/src/test/java/org/springframework/cloud/circuitbreaker/resilience4j/BulkheadConfigurationTests.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+	@Test
```
Comment (OathMeadow) [inline spring-cloud-circuitbreaker-resilience4j/src/test/java/org/springframework/cloud/circuitbreaker/resilience4j/BulkheadConfigurationTests.java:36]: Do we want to test these?
What is expected?

`Resilience4jBulkheadProvider::configureDefault`
`Resilience4jBulkheadProvider::configure`
`Resilience4jBulkheadProvider::addBulkheadCustomizer`
`Resilience4jBulkheadProvider::addThreadPoolBulkheadCustomizer`

Comment (OathMeadow) [reply spring-cloud-circuitbreaker-resilience4j/src/test/java/org/springframework/cloud/circuitbreaker/resilience4j/BulkheadConfigurationTests.java:36]: Example:

`Resilience4jBulkheadProvider::configureDefault` vs `resilience4j.threadPoolBulkHead.configs.default.queueCapacity=30`

In my opinion, when `Resilience4jBulkheadProvider::configureDefault` is used, then it take precedence over property. Perhaps same with `Resilience4jBulkheadProvider::configure` ?

Comment (ryanjbaxter) [reply spring-cloud-circuitbreaker-resilience4j/src/test/java/org/springframework/cloud/circuitbreaker/resilience4j/BulkheadConfigurationTests.java:37]: Resilience4jBulkheadProvider::configureDefault, Resilience4jBulkheadProvider::addBulkheadCustomizer, Resilience4jBulkheadProvider::addThreadPoolBulkheadCustomizer are used in Resilience4JBulkheadIntegrationTest and  and Resilience4JBulkheadAndTimeLimiterIntegrationTest.

I agree on the configureDefault and I added some tests for that.

**Review by ryanjbaxter — COMMENTED** [review]
Thanks, made some changes based on your feedback

File: spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4jBulkheadProvider.java [hunk 4]
```diff
@@ -143,11 +143,13 @@ private <T> Supplier<CompletionStage<T>> decorateBulkhead(final String id, final
 
 	private Resilience4jBulkheadConfigurationBuilder.BulkheadConfiguration getConfiguration(String id) {
 		Resilience4jBulkheadConfigurationBuilder builder = new Resilience4jBulkheadConfigurationBuilder();
+		Resilience4jBulkheadConfigurationBuilder.BulkheadConfiguration defaultConfiguration = this.defaultConfiguration
```
Comment (OathMeadow) [inline spring-cloud-circuitbreaker-resilience4j/src/main/java/org/springframework/cloud/circuitbreaker/resilience4j/Resilience4jBulkheadProvider.java:146]: This is just a minor suggestion really.
The method `configureDefault(..)` allows to set the variable `defaultConfiguration` to `null`.

**Review by OathMeadow — APPROVED** [review]

**Pull request merged**